src/EditorFeatures/Core.Wpf/StringIndentation/StringIndentationTag.cs

```diff
@@ -31,5 +37,35 @@ internal class StringIndentationTag : BrushTag
             var brush = view.VisualElement.TryFindResource("outlining.verticalrule.foreground") as SolidColorBrush;
             return brush?.Color;
         }
+
+        public override int GetHashCode()
+            => throw ExceptionUtilities.Unreachable();
+
+        public override bool Equals(object? obj)
+            => Equals(obj as StringIndentationTag);
+
+        public bool Equals(StringIndentationTag? other)
+        {
+            if (other is null)
+                return false;
+
+            if (this.OrderedHoleSpans.Length != other.OrderedHoleSpans.Length)
+                return false;
+
+            for (int i = 0, n = this.OrderedHoleSpans.Length; i < n; i++)
+            {
+                if (!SpanEquals(this.OrderedHoleSpans[i], other.OrderedHoleSpans[i]))
+                    return false;
+            }
+
+            return true;
+        }
+
+        private bool SpanEquals(SnapshotSpan snapshotSpan1, SnapshotSpan snapshotSpan2)
+        {
+            // Map span1 over to span2
+            var span1 = snapshotSpan1.TranslateTo(snapshotSpan2.Snapshot, _provider.SpanTrackingMode);
+            return span1.Span == snapshotSpan2.Span;
+        }
```

Comment:
more complex tags have to normalize themselves to do the equality comparison properly.